src/unix/os390.c

```diff
@@ -665,23 +668,122 @@ int uv__io_check_fd(uv_loop_t* loop, int fd) {
 
 
 void uv__fs_event_close(uv_fs_event_t* handle) {
-  UNREACHABLE();
+  uv__os390_epoll* ep;
+  
+  ep = handle->loop->ep;
+  if (epoll_ctl(ep, UV__EPOLL_CTL_MSGQD, handle->loop->fs_event_msg_queue, NULL))
+    abort();
+
+  /* No message events left. So close the queue */
+  if (ep->num_fs_events == 0) {
+    msgctl(handle->loop->fs_event_msg_queue, IPC_RMID, NULL);
+    handle->loop->fs_event_msg_queue = -1;
+  }
 }
 
 
 int uv_fs_event_init(uv_loop_t* loop, uv_fs_event_t* handle) {
-  return -ENOSYS;
+  // Create message queue
+  struct epoll_event e;
+
+  if (loop->fs_event_msg_queue == -1)
+    loop->fs_event_msg_queue = msgget(IPC_PRIVATE, 0622 | IPC_CREAT);
+
+  if (loop->fs_event_msg_queue == -1)
+    abort();
+
+  e.fd = loop->fs_event_msg_queue;
+  e.events = POLLIN;
+  if (epoll_ctl(loop->ep, UV__EPOLL_CTL_MSGQA, loop->fs_event_msg_queue, &e))
+    abort();
+  // Register handle
+  uv__handle_init(loop, (uv_handle_t*)handle, UV_FS_EVENT);
+  return 0;
 }
 
 
 int uv_fs_event_start(uv_fs_event_t* handle, uv_fs_event_cb cb,
                       const char* filename, unsigned int flags) {
-  return -ENOSYS;
+  _RFIS reg_struct;
+  int rc;
+
+  if (handle->loop->fs_event_msg_queue == -1)
+    abort();
```

Comment:
assert?  Hitting the abort would be a libuv bug, wouldn't it, not a runtime error?